modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/hooks/useVariants.es.js

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import React, {createContext, useContext, useState} from 'react';
+
+const VariantsContext = createContext({
+	components: {},
+	setComponents: () => {},
+});
+
+export const VariantsProvider = ({children, components: initialComponents}) => {
+	const [components, setComponents] = useState(initialComponents);
+
+	return (
+		<VariantsContext.Provider value={{components, setComponents}}>
+			{children}
+		</VariantsContext.Provider>
+	);
+};
+
+export const useVariants = () => {
+	return useContext(VariantsContext);
+};
+
+VariantsContext.displayName = 'VariantsContext';
```

Comment:
@diegonvs would avoid creating a `VariantProvider` and a `useVariants` for this case, make it called only in the `Renderer` component and being used in the `Layout` component, we don’t want it to be used elsewhere.

For example:

- Call `<VariantsContext.Provider value={Components}>` directly in `Renderer`
- Use the `VariantsContext` directly in the `Layout` component.

Pseudocode
```jsx
export const Layout = ({
	components: Components = DefaultVariant,
}) => {
	const defaultComponents = useContext(VariantsContext);
	
	const Components = components ?? defaultComponents;
};
```

Reply:
> @diegonvs would avoid creating a VariantProvider and a useVariants for this case, make it called only in the Renderer component and being used in the Layout component, we don’t want it to be used elsewhere.

We are needing the `components` information on Renderer component and on the [FieldSet(field-type) component](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/FieldSet/FieldSet.es.js#L65)

Reply:
@diegonvs yeah, once we add this to the Layout directly, we no longer need to add it to the FieldSet.